BUG: Test suite leaks file descriptors (`OSError: [Errno 24] Too many open files` on PyPy3.10)

### Describe your issue.

When running the complete test suite on PyPy3.10, I'm getting lots of test failurs (almost 1000) directly related to errors such as:

```pytb
ERROR scipy/fft/_pocketfft/tests/test_real_transforms.py::test_idst_definition[13-4-int] - OSError: [Errno 24] Too many open files: '/v
ar/tmp/portage/dev-python/scipy-1.11.3-r1/work/scipy-1.11.3-pypy3/install/usr/lib/pypy...
```

After investigating the problem, it seems that `get_reference_data()` in `scipy/fft/_pocketfft/tests/test_real_transforms.py` is calling `np.load()` a few times. The returned objects hold an open file handle. Since they are neither explicitly `.close()`d nor used via a context manager, the file remains open until GC cleans the object up. The GC in PyPy is less aggressive, therefore it becomes possible for the fds to remain open until hitting the system limit.

FWICS `get_reference_data()` attempts to cache the data in `globals()`. However, this doesn't seem to work at all — neither in PyPy3.10, nor in CPython 3.11. If I add a `raise ValueError(1)` below `if ref is not None:`, it never triggers.

The same seems to apply to other uses of `np.load()` across the codebase but I suppose they aren't called as often as in this particular test, so they're not a critical problem right now.

### Reproducing Code Example

```python
pypy3.10 -m pip install scipy pytest hypothesis
# the issue is easier to reproduce with lower fd limit
ulimit -n 64
pypy3.10 -m pytest --pyargs scipy.fft._pocketfft.tests.test_real_transforms
```


### Error message

```shell
_________________________________________ ERROR at setup of TestDCT.test_definition[0-2-int] __________________________________________

request = <SubRequest 'fftwdata_size' for <Function test_definition[0-2-int]>>

>   ???

.venv/lib/pypy3.10/site-packages/scipy/fft/_pocketfft/tests/test_real_transforms.py:70: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
.venv/lib/pypy3.10/site-packages/scipy/fft/_pocketfft/tests/test_real_transforms.py:34: in get_reference_data
    ???
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

file = '/tmp/.venv/lib/pypy3.10/site-packages/scipy/fft/_pocketfft/tests/../../../fftpack/tests/test.npz', mmap_mode = None
allow_pickle = False, fix_imports = True, encoding = 'ASCII'

>   ???
E   OSError: [Errno 24] Too many open files: '/tmp/.venv/lib/pypy3.10/site-packages/scipy/fft/_pocketfft/tests/../../../fftpack/tests/test.npz'

.venv/lib/pypy3.10/site-packages/numpy/lib/npyio.py:427: OSError
```


### SciPy/NumPy/Python version and system information

```shell
1.11.3 1.26.2 sys.version_info(major=3, minor=10, micro=13, releaselevel='final', serial=0)
/tmp/.venv/lib/pypy3.10/site-packages/scipy/__config__.py:146: UserWarning: Install `pyyaml` for better output
  warnings.warn("Install `pyyaml` for better output", stacklevel=1)
{
  "Compilers": {
    "c": {
      "name": "gcc",
      "linker": "ld.bfd",
      "version": "13.2.1",
      "commands": "ccache, cc"
    },
    "cython": {
      "name": "cython",
      "linker": "cython",
      "version": "0.29.36",
      "commands": "cython"
    },
    "c++": {
      "name": "gcc",
      "linker": "ld.bfd",
      "version": "13.2.1",
      "commands": "ccache, c++"
    },
    "fortran": {
      "name": "gcc",
      "linker": "ld.bfd",
      "version": "13.2.1",
      "commands": "gfortran"
    },
    "pythran": {
      "version": "0.14.0",
      "include directory": "../../../pip-build-env-120xme0q/overlay/lib/pypy3.10/site-packages/pythran"
    }
  },
  "Machine Information": {
    "host": {
      "cpu": "x86_64",
      "family": "x86_64",
      "endian": "little",
      "system": "linux"
    },
    "build": {
      "cpu": "x86_64",
      "family": "x86_64",
      "endian": "little",
      "system": "linux"
    },
    "cross-compiled": false
  },
  "Build Dependencies": {
    "blas": {
      "name": "openblas",
      "found": true,
      "version": "0.3.25",
      "detection method": "pkgconfig",
      "include directory": "/usr/include/openblas",
      "lib directory": "/usr/lib64",
      "openblas configuration": "USE_64BITINT= DYNAMIC_ARCH= DYNAMIC_OLDER= NO_CBLAS= NO_LAPACK= NO_LAPACKE= NO_AFFINITY=1 USE_OPENMP=1 ZEN MAX_THREADS=64",
      "pc file directory": "/usr/lib64/pkgconfig"
    },
    "lapack": {
      "name": "openblas",
      "found": true,
      "version": "0.3.25",
      "detection method": "pkgconfig",
      "include directory": "/usr/include/openblas",
      "lib directory": "/usr/lib64",
      "openblas configuration": "USE_64BITINT= DYNAMIC_ARCH= DYNAMIC_OLDER= NO_CBLAS= NO_LAPACK= NO_LAPACKE= NO_AFFINITY=1 USE_OPENMP=1 ZEN MAX_THREADS=64",
      "pc file directory": "/usr/lib64/pkgconfig"
    },
    "pybind11": {
      "name": "pybind11",
      "version": "2.11.1",
      "detection method": "pkgconfig",
      "include directory": "/usr/include"
    }
  },
  "Python Information": {
    "path": "/tmp/.venv/bin/pypy3.10",
    "version": "3.10"
  }
}
```

Yes, I can. On top of e7c6a0d1ce504c528ab00d19ddac6bdb1ae1cb36:

```
💻  ninja -C /home/mgorny/git/scipy/build -j6
ninja: Entering directory `/home/mgorny/git/scipy/build'
[1/82] Generating scipy/linalg/_dummy_cython_blas.pxd with a custom command
[2/46] Generating scipy/special/_dummy_cython_special.pxd with a custom command
[3/31] Generating scipy/stats/_dummy_stats_gen_pyx with a custom command
[4/4] Generating scipy/generate-version with a custom command
Build OK
💻  meson install -C build --only-changed
Installing, see meson-install.log...
Installation OK
SciPy from development installed path at: /home/mgorny/git/scipy/build-install/lib/pypy3.10/site-packages
Running tests for scipy version:1.13.0.dev0+1134.e7c6a0d, installed at:/home/mgorny/git/scipy/build-install/lib/pypy3.10/site-packages/scipy
============================= test session starts ==============================
platform linux -- Python 3.10.13[pypy-7.3.14-final], pytest-7.4.4, pluggy-1.3.0
rootdir: /home/mgorny/git/scipy
configfile: pytest.ini
plugins: hypothesis-6.92.5
collected 1948 items

scipy/fft/_pocketfft/tests/test_real_transforms.py ..................FEE [  1%]
EEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEE [  4%]
EEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEE [  8%]
EEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEE.. [ 12%]
........................................................................ [ 15%]
.................................EEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEE [ 19%]
EEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEE [ 23%]
EEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEE.. [ 26%]
....EEEEE......EEEEEEE......EEEEEEEE......EEEEEEEE......EEEEEEEE......EE [ 30%]
EEEEE......EEEEEEEEE......EEEEEEEE......EEEEEEEEE......EEEEEEEE......EEE [ 34%]
EEEEEE......EEEEEEEE......EEEEEEEEE......EEEEEEEEE......EEEEEEEEE......E [ 38%]
EEEEEEEE......EEEEEEEEE......EEEEEEEEE......EEEEEEEEE......EEEEEEEEE.... [ 41%]
..EEEEEEEEE......EEEEEEEEE......EEEEEEEEE......EEEEEEEEE......EEEEEEEEE. [ 45%]
.....EEEEEEEEE......EEEEEEEEE............EEEEEEEE.................EEE... [ 49%]
.........EEEEE......FEE......EEEEEEEE......EEEEEEEE......EEEEEEEEE...... [ 52%]
EEEEEEEE......EEEEEEEEE......EEEEEEEEE......EEEEEEEEE......EEEEEEEEE.... [ 56%]
..EEEEEEEEE......EEEEEEEEE......EEEEEEEEE......EEEEEEEEE......EEEEEEEEE. [ 60%]
.....EEEEEEEEE......EEE................................................. [ 63%]
........................................................................ [ 67%]
........................................................................ [ 71%]
........................................................................ [ 75%]
........................................................................ [ 78%]
........................................................................ [ 82%]
........................................................................ [ 86%]
........................................................................ [ 89%]
........................................................................ [ 93%]
........................................................................ [ 97%]
.......................................................                  [100%]

==================================== ERRORS ====================================
______________ ERROR at setup of TestDCT.test_definition[0-2-int] ______________
scipy/fft/_pocketfft/tests/test_real_transforms.py:70: in fftwdata_size
    ???
        request    = <SubRequest 'fftwdata_size' for <Function test_definition[0-2-int]>>
scipy/fft/_pocketfft/tests/test_real_transforms.py:34: in get_reference_data
    ???
        ref        = None
../../../../.venv/lib/pypy3.10/site-packages/numpy/lib/npyio.py:427: in load
    ???
E   OSError: [Errno 24] Too many open files: '/home/mgorny/git/scipy/build-install/lib/pypy3.10/site-packages/scipy/fft/_pocketfft/tests/../../../fftpack/tests/test.npz'
        allow_pickle = False
        encoding   = 'ASCII'
        file       = '/home/mgorny/git/scipy/build-install/lib/pypy3.10/site-packages/scipy/fft/_pocketfft/tests/../../../fftpack/tests/test.npz'
        fix_imports = True
        max_header_size = 10000
        mmap_mode  = None
        pickle_kwargs = {'encoding': 'ASCII', 'fix_imports': True}
        stack      = <contextlib.ExitStack object at 0x000055e4790dc250>
__________ ERROR at setup of TestDCT.test_definition[0-3-longdouble] ___________
scipy/fft/_pocketfft/tests/test_real_transforms.py:70: in fftwdata_size
    ???
        request    = <SubRequest 'fftwdata_size' for <Function test_definition[0-3-longdouble]>>
scipy/fft/_pocketfft/tests/test_real_transforms.py:34: in get_reference_data
    ???
        ref        = None
../../../../.venv/lib/pypy3.10/site-packages/numpy/lib/npyio.py:427: in load
    ???
E   OSError: [Errno 24] Too many open files: '/home/mgorny/git/scipy/build-install/lib/pypy3.10/site-packages/scipy/fft/_pocketfft/tests/../../../fftpack/tests/test.npz'
        allow_pickle = False
        encoding   = 'ASCII'
        file       = '/home/mgorny/git/scipy/build-install/lib/pypy3.10/site-packages/scipy/fft/_pocketfft/tests/../../../fftpack/tests/test.npz'
        fix_imports = True
        max_header_size = 10000
        mmap_mode  = None
        pickle_kwargs = {'encoding': 'ASCII', 'fix_imports': True}
        stack      = <contextlib.ExitStack object at 0x000055e4790ea6e8>
____________ ERROR at setup of TestDCT.test_definition[0-3-float64] ____________
scipy/fft/_pocketfft/tests/test_real_transforms.py:70: in fftwdata_size
    ???
        request    = <SubRequest 'fftwdata_size' for <Function test_definition[0-3-float64]>>
scipy/fft/_pocketfft/tests/test_real_transforms.py:34: in get_reference_data
    ???
        ref        = None
../../../../.venv/lib/pypy3.10/site-packages/numpy/lib/npyio.py:427: in load
    ???
E   OSError: [Errno 24] Too many open files: '/home/mgorny/git/scipy/build-install/lib/pypy3.10/site-packages/scipy/fft/_pocketfft/tests/../../../fftpack/tests/test.npz'
        allow_pickle = False
        encoding   = 'ASCII'
        file       = '/home/mgorny/git/scipy/build-install/lib/pypy3.10/site-packages/scipy/fft/_pocketfft/tests/../../../fftpack/tests/test.npz'
        fix_imports = True
        max_header_size = 10000
        mmap_mode  = None
        pickle_kwargs = {'encoding': 'ASCII', 'fix_imports': True}
        stack      = <contextlib.ExitStack object at 0x000055e4790fc288>
[…]
```

--- a/scipy/fft/_pocketfft/tests/test_real_transforms.py
+++ b/scipy/fft/_pocketfft/tests/test_real_transforms.py
@@ -25,11 +25,8 @@ def is_longdouble_binary_compatible():
         return False
 
 
-def get_reference_data():
-    ref = getattr(globals(), '__reference_data', None)
-    if ref is not None:
-        return ref
-
+@pytest.fixture(scope="module")
+def reference_data():
     # Matlab reference data
     MDATA = np.load(join(fftpack_test_dir, 'test.npz'))
     X = [MDATA['x%d' % i] for i in range(MDATA_COUNT)]
@@ -61,51 +58,55 @@ def get_reference_data():
         'Y': Y
     }
 
-    globals()['__reference_data'] = ref
-    return ref
+    yield ref
+
+    if is_longdouble_binary_compatible():
+        FFTWDATA_LONGDOUBLE.close()
+    FFTWDATA_SINGLE.close()
+    FFTWDATA_DOUBLE.close()
+    MDATA.close()
 
 
 @pytest.fixture(params=range(FFTWDATA_COUNT))
-def fftwdata_size(request):
-    return get_reference_data()['FFTWDATA_SIZES'][request.param]
+def fftwdata_size(request, reference_data):
+    return reference_data['FFTWDATA_SIZES'][request.param]
 
 @pytest.fixture(params=range(MDATA_COUNT))
-def mdata_x(request):
-    return get_reference_data()['X'][request.param]
+def mdata_x(request, reference_data):
+    return reference_data['X'][request.param]
 
 
 @pytest.fixture(params=range(MDATA_COUNT))
-def mdata_xy(request):
-    ref = get_reference_data()
-    y = ref['Y'][request.param]
-    x = ref['X'][request.param]
+def mdata_xy(request, reference_data):
+    y = reference_data['Y'][request.param]
+    x = reference_data['X'][request.param]
     return x, y
 
 
-def fftw_dct_ref(type, size, dt):
+def fftw_dct_ref(type, size, dt, reference_data):
     x = np.linspace(0, size-1, size).astype(dt)
     dt = np.result_type(np.float32, dt)
     if dt == np.float64:
-        data = get_reference_data()['FFTWDATA_DOUBLE']
+        data = reference_data['FFTWDATA_DOUBLE']
     elif dt == np.float32:
-        data = get_reference_data()['FFTWDATA_SINGLE']
+        data = reference_data['FFTWDATA_SINGLE']
     elif dt == np.longdouble:
-        data = get_reference_data()['FFTWDATA_LONGDOUBLE']
+        data = reference_data['FFTWDATA_LONGDOUBLE']
     else:
         raise ValueError()
     y = (data['dct_%d_%d' % (type, size)]).astype(dt)
     return x, y, dt
 
 
-def fftw_dst_ref(type, size, dt):
+def fftw_dst_ref(type, size, dt, reference_data):
     x = np.linspace(0, size-1, size).astype(dt)
     dt = np.result_type(np.float32, dt)
     if dt == np.float64:
-        data = get_reference_data()['FFTWDATA_DOUBLE']
+        data = reference_data['FFTWDATA_DOUBLE']
     elif dt == np.float32:
-        data = get_reference_data()['FFTWDATA_SINGLE']
+        data = reference_data['FFTWDATA_SINGLE']
     elif dt == np.longdouble:
-        data = get_reference_data()['FFTWDATA_LONGDOUBLE']
+        data = reference_data['FFTWDATA_LONGDOUBLE']
     else:
         raise ValueError()
     y = (data['dst_%d_%d' % (type, size)]).astype(dt)
@@ -265,8 +266,8 @@ for k,v in dec_map.copy().items():
 @pytest.mark.parametrize('rdt', [np.longdouble, np.float64, np.float32, int])
 @pytest.mark.parametrize('type', [1, 2, 3, 4])
 class TestDCT:
-    def test_definition(self, rdt, type, fftwdata_size):
-        x, yr, dt = fftw_dct_ref(type, fftwdata_size, rdt)
+    def test_definition(self, rdt, type, fftwdata_size, reference_data):
+        x, yr, dt = fftw_dct_ref(type, fftwdata_size, rdt, reference_data)
         y = dct(x, type=type)
         assert_equal(y.dtype, dt)
         dec = dec_map[(dct, rdt, type)]
@@ -340,8 +341,8 @@ def test_dct4_definition_ortho(mdata_x, rdt):
 
 @pytest.mark.parametrize('rdt', [np.longdouble, np.float64, np.float32, int])
 @pytest.mark.parametrize('type', [1, 2, 3, 4])
-def test_idct_definition(fftwdata_size, rdt, type):
-    xr, yr, dt = fftw_dct_ref(type, fftwdata_size, rdt)
+def test_idct_definition(fftwdata_size, rdt, type, reference_data):
+    xr, yr, dt = fftw_dct_ref(type, fftwdata_size, rdt, reference_data)
     x = idct(yr, type=type)
     dec = dec_map[(idct, rdt, type)]
     assert_equal(x.dtype, dt)
@@ -350,8 +351,8 @@ def test_idct_definition(fftwdata_size, rdt, type):
 
 @pytest.mark.parametrize('rdt', [np.longdouble, np.float64, np.float32, int])
 @pytest.mark.parametrize('type', [1, 2, 3, 4])
-def test_definition(fftwdata_size, rdt, type):
-    xr, yr, dt = fftw_dst_ref(type, fftwdata_size, rdt)
+def test_definition(fftwdata_size, rdt, type, reference_data):
+    xr, yr, dt = fftw_dst_ref(type, fftwdata_size, rdt, reference_data)
     y = dst(xr, type=type)
     dec = dec_map[(dst, rdt, type)]
     assert_equal(y.dtype, dt)
@@ -384,8 +385,8 @@ def test_dst4_definition_ortho(rdt, mdata_x):
 
 @pytest.mark.parametrize('rdt', [np.longdouble, np.float64, np.float32, int])
 @pytest.mark.parametrize('type', [1, 2, 3, 4])
-def test_idst_definition(fftwdata_size, rdt, type):
-    xr, yr, dt = fftw_dst_ref(type, fftwdata_size, rdt)
+def test_idst_definition(fftwdata_size, rdt, type, reference_data):
+    xr, yr, dt = fftw_dst_ref(type, fftwdata_size, rdt, reference_data)
     x = idst(yr, type=type)
     dec = dec_map[(idst, rdt, type)]
     assert_equal(x.dtype, dt)

